pkg/clusterapi/aws/actuator.go

```diff
@@ -367,35 +340,114 @@ func (a *Actuator) DeleteMachine(machine *clusterv1.Machine) error {
+// for details that may have changed. (IPs and hostnames) We do not currently support making any
+// changes to actual machines in AWS. Instead these will be replaced via MachineDeployments.
+func (a *Actuator) Update(cluster *clusterv1.Cluster, machine *clusterv1.Machine) error {
+	mLog := clustoplog.WithMachine(a.logger, machine)
+	mLog.Debugf("updating machine")
+
+	clusterSpec, err := controller.ClusterSpecFromClusterAPI(cluster)
 	if err != nil {
-		// If the instance no longer exists, we need to make sure we don't infinitely
-		// block deletion of the Machine in etcd.
-		if awsErr, ok := err.(awserr.Error); ok {
-			if awsErr.Code() == ec2InstanceIDNotFoundCode {
-				mLog.WithField("instanceID", instanceID).Warnf("AWS instance no longer exists")
-				return nil
+		return err
+	}
+
+	if clusterSpec.Hardware.AWS == nil {
+		return fmt.Errorf("Cluster does not contain an AWS hardware spec")
+	}
+
+	coMachineSetSpec, err := controller.MachineSetSpecFromClusterAPIMachineSpec(&machine.Spec)
+	if err != nil {
+		return err
+	}
+
+	region := clusterSpec.Hardware.AWS.Region
+	mLog.WithField("region", region).Debugf("obtaining EC2 client for region")
+	client, err := a.ec2Client(coMachineSetSpec, machine.Namespace, region)
+	if err != nil {
+		return fmt.Errorf("unable to obtain EC2 client: %v", err)
+	}
+
+	instances, err := a.getRunningInstances(machine, client)
+	if err != nil {
+		return err
+	}
+
+	if len(instances) == 0 {
+		// Parent controller should prevent this from ever happening by calling Exists and then Create.
+		return fmt.Errorf("attempted to update machine but no instances found")
+	}
+
+	// In very unusual circumstances, there could be more than one machine running matching this
+	// machine name and cluster ID. In this scenario we will keep the newest, and delete all others.
+	var newestInstance *ec2.Instance
+	if len(instances) > 1 {
+
+		// Determine which instance is the newest:
+		for _, instance := range instances {
+			mLog.WithFields(log.Fields{
+				"instanceID": *instance.InstanceId,
+				"state":      *instance.State.Name,
+				"launchTime": *instance.LaunchTime,
+			}).Warn("found multiple instances for machine in pending/running state")
+
+			if newestInstance == nil {
+				newestInstance = instance
+				continue
 			}
+			newestInstance = chooseNewestInstance(newestInstance, instance, mLog)
 		}
-		return fmt.Errorf("error terminating instance %q: %v", instanceID, err)
+
+		// Cleanup all the older instances:
+		terminateInstanceIDs := []*string{}
+		for _, instance := range instances {
+			if *instance.InstanceId != *newestInstance.InstanceId {
+				terminateInstanceIDs = append(terminateInstanceIDs, instance.InstanceId)
+			}
+		}
+		err := a.terminateInstances(client, terminateInstanceIDs, mLog)
+		if err != nil {
+			return err
+		}
+	} else {
+		newestInstance = instances[0]
+	}
+
+	mLog = mLog.WithField("instanceID", *newestInstance.InstanceId)
+	mLog.Debug("instance found")
+
+	instance := instances[0]
```

Comment:
Should this be `instance := newestInstance`?

Reply:
Good catch, ty.